### PR TITLE
Make Flaky into an ExceptionGroup

### LINKED ISSUE #4040: Change Flaky to be an ExceptionGroup
`Flaky` can be caused by one *or two different* failures. Changing it to inherit `BaseExceptionGroup` will make it possible to improve reporting of the underlying exceptions.

_see https://github.com/HypothesisWorks/hypothesis/pull/4023#discussion_r1672186714_

### PR DESCRIPTION
**Pull request by jobh**
This PR makes `Flaky` into an `ExceptionGroup` to improve presentation of flaky failures. Closes #4040.

As usual, *There Were Complications*. Basically because an `ExceptionGroup` is not allowed to have an empty group; so a straightforward replacement is not possible.

On the plus side, I think the reworked `Flaky` is more expressive by differentiating two different kinds of flakiness (replay inconsistencies vs. inconsistent data generation), and it does improve reporting of flaky errors detected by the engine by adding full exceptions on the way out.

```
HypothesisException (via _Trimmable)
    |
    +Flaky              Base class for all flakiness (not currently raised anywhere).
       |
       + FlakyData      Raised to user. Conjecture detects inconsistent data generation.
       |
       + FlakyReplay    Internal. Conjecture detects flaky replay failure, but has only InterestingOrigins.
       |
       + FlakyFailure   Raised to user. ExceptionGroup, possibly constructed from a FlakyReplay.
```

A simpler flat alternative is this, with the only drawback being that `InconsistentData` (renamed from FlakyData) is no longer `Flaky`. I think I actually prefer this version, but I don't know how you feel about this drawback — is it even acceptable for a minor revision, do users ever catch `Flaky`? — so the implemented version is the one above.

```
HypothesisException (via _Trimmable)
    |
    + Flaky                Raised to user. ExceptionGroup, possibly constructed from a FlakyReplay.
    |
    + FlakyReplay          Internal. Conjecture detects flaky replay failure, but has only InterestingOrigins.
    |
    + InconsistentData     Raised to user. Conjecture detects inconsistent data generation.
```

### CODE REVIEW

**Comment by tybug** [comment]
>  is it even acceptable for a minor revision, do users ever catch Flaky?

Looks pretty rare, with the majority being schemathesis: https://github.com/search?q=%22except+Flaky%3A%22+OR+%22except+Flaky+as%22+language%3Apython&type=code

That said, I definitely have some local research code that catches Flaky...but I knew I was flying close to the sun with that one.

**Comment by Zac-HD** [comment]
Seems rare, but while I share your preference for the flat version I think we should continue to support `Flaky` - power-users and researchers are still users, and I'm generally comfortable taking on some more maintainece burden to make using Hypothesis delightful 🙂

File: hypothesis-python/src/hypothesis/errors.py [hunk 1]
```diff
@@ -67,6 +102,19 @@ class Flaky(_Trimmable):
            don't do that and testing those instead.
     """
 
+    def __new__(cls, msg, group):
+        # The Exception mixin forces this an ExceptionGroup (only accepting
+        # Exceptions, not BaseException). Usually BaseException is raised
+        # directly and will hence not be part of a FlakyFailure, but I'm not
+        # sure this assumption holds everywhere. So wrap any BaseExceptions.
```
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/errors.py:109]: Let's just inherit from `BaseExceptionGroup` to get around that?

**Comment by jobh** [comment]
So be it 😇 Ready for review then! No rush - I'm starting my summer vacation today, and will be on and off depending on weather. Hopefully mostly off 😁

**Review by Zac-HD — COMMENTED** [review]
I want to play around and compare the before-and-after reporting for some examples before we merge, but this looks great so far 😁

File: hypothesis-python/src/hypothesis/errors.py (hunk 1 above)
Comment (jobh) [reply hypothesis-python/src/hypothesis/errors.py:109]: Unfortunately, the HypothesisException mixing makes it non-base even so. On mobile now, so cant find the ref, but mixing in Exception defines ExceptionGroup.

File: hypothesis-python/src/hypothesis/errors.py [hunk 2]
```diff
@@ -52,9 +54,42 @@ class Unsatisfiable(_Trimmable):
...
+    a FlakyFailure exception group for final presentation.
+    """
+
+    def __init__(self, reason, interesting_origins=None):
+        super().__init__(reason)
+        self.reason = reason
+        self._interesting_origins = interesting_origins
+
+
+class FlakyData(Flaky):
```
[the remark this answers is not in the document]
Comment (jobh) [reply hypothesis-python/src/hypothesis/errors.py:75]: Naming is hard! I went through FlakyGeneration plus some variants of InconsistentX  and IndetermisticX, and not at all sure I settled for the best.

[the remark this answers is not in the document]
Comment (jobh) [reply hypothesis-python/src/hypothesis/errors.py:75]: FlakyStrategy?

[the remark this answers is not in the document]
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/errors.py:75]: I do like `FlakyReplay` as a concise and evocative name; for the other I've gone with `FlakyStrategyDefinition` which is more verbose but does seem to point to the thing-which-is-flaky.

**Review by Zac-HD — APPROVED** [review]
Played around a bit more and I'm very happy with this; I've also pulled out a cross-backend-specific thing from https://github.com/HypothesisWorks/hypothesis/pull/4034 which will help us and our users debug such issues.

Thanks again, @jobh!

**Pull request merged**